engine/src/test/scala/cromwell/jobstore/JobStoreWriterSpec.scala

```diff
@@ -0,0 +1,97 @@
+package cromwell.jobstore
+
+import akka.actor.ActorRef
+import cromwell.CromwellTestkitSpec
+import cromwell.core.WorkflowId
+import cromwell.jobstore.JobStoreService.{JobStoreWriteSuccess, RegisterJobCompleted, RegisterWorkflowCompleted}
+import org.scalatest.{BeforeAndAfter, Matchers}
+
+import scala.concurrent.duration._
+import language.postfixOps
+import scala.concurrent.{ExecutionContext, Future}
+
+class JobStoreWriterSpec extends CromwellTestkitSpec with Matchers with BeforeAndAfter {
+  
+  var database: WriteCountingJobStoreDatabase = _
+  var jobStoreWriter: ActorRef = _
+  var workflowId: WorkflowId = _
+  val successResult: JobResult = JobResultSuccess(Some(0), Map.empty)
+
+  before {
+    database = WriteCountingJobStoreDatabase.makeNew
+    jobStoreWriter = system.actorOf(JobStoreWriterActor.props(database))
+    workflowId = WorkflowId.randomId()
+  }
+
+  private def registerCompletions(attempts: Int): Unit = {
+    0 until attempts foreach { a => jobStoreWriter ! RegisterJobCompleted(jobKey(attempt = a), successResult) }
+  }
+
+  private def jobKey(attempt: Int): JobStoreKey = JobStoreKey(workflowId, s"call.fqn", None, attempt)
+
+  private def assertWriteSuccess(key: JobStoreKey, result: JobResult): Unit = {
+    key.workflowId shouldBe workflowId
+    key.callFqn shouldBe "call.fqn"
+    key.index shouldBe None
+    result shouldBe successResult
+  }
+
+  private def assertDb(totalWritesCalled: Int, jobCompletionsRecorded: Int, workflowCompletionsRecorded: Int): Unit = {
+    database.totalWritesCalled shouldBe totalWritesCalled
+    database.jobCompletionsRecorded shouldBe jobCompletionsRecorded
+    database.workflowCompletionsRecorded shouldBe workflowCompletionsRecorded
+  }
+
+  private def assertReceived(allowWorkflowCompleted: Boolean): Unit = {
+    val received = receiveN(3, 10 seconds)
```

Comment:
Instead of a hardcoded `3` and the `allowWorkflowCompleted`, it'd be nice to have two parameters: `expectedJobCompletionMessages` and `expectedWorkflowCompletionMessages`